### ISSUE TITLE
[Question] Computing the predictive posterior mean 

### ISSUE THREAD

**Issue by jodablei** [cornellius-gp/gpytorch#682]
Hi,

I am not sure if I encountered a bug when computing the predictive posterior mean (exact gp). Basically, the problem seems to be in DefaultPredictionStrategy.exact_predictive_mean when test_train_covar.dim() == 2. In this case, torch.addmv is used, but the vectors test_mean and self.mean_cache are 2D tensors (w/ singleton dimension). However, the method is expecting 1D tensors. Using .squeeze can fix this issue, but I am wondering if I am the only one that has encountered this problem. Even the 01_Simple_GP_Regression example throws this error when making predictions.

**Comment by jacobrgardner** [comment]
@jodablei Everything seems to be working for me e.g. in the simple example notebook. Do you have specific code that produces the issue? What version of GPyTorch are you running on?

**Comment by jodablei** [comment]
Apparently the error was caused by a custom mean function I wrote. It's been fixed. 

A bit unrelated to this topic, but I have some additional questions:

1. Can you shed some light as to why GPyTorch is redefining the constraint objects, instead of using those in torch.distributions.constraints + torch.distributions.biject_to? I also think the transform in gpytorch.constraints.Interval might have a bug. Shouldn't the scale be (upper_bound - lower_bound), instead of upper_bound ?

2. Is the _bufferize_attributes call necessary when implementing new distributions (to be used as priors)? Will there be any bad side effects if I don't? I tried to wrap torch.distributions.LogNormal, but I was getting an error in _bufferize_attributes, specifically when trying to delete the attributes 'loc' and 'scale'. It would be useful to include an utility that wraps all the existing torch.distributions. 

Thanks in advance

**Comment by Balandat** [comment]
re 2.: This is done so that the priors can easily be moved to a different device/dtype. Say you have a model `m` on the CPU with some prior - without the bufferize logic in the constructor, you couldn't do `m.to(device=torch.device("cuda"))` without running into device/dtype incompatibilities. Using buffers avoids having to re-instantiate the distribution with the tensor on the new device / with the new dtype.

**Comment by jacobrgardner** [comment]
@jodablei 

Regarding 1: The reason we do this is because the constraints in `torch.distributions` all use `torch.exp` as the mapping to ensure positiveness. In our experiments, this is significantly worse than using `torch.nn.functional.softplus`. 

When using `exp`, small perturbations in the raw parameters result in very large perturbations in the true parameters, which we found makes optimization very unstable.
